src/models/room-state.ts

```diff
@@ -493,22 +493,24 @@ export class RoomState extends TypedEventEmitter<EmittedEvents, EventHandlerMap>
             }
         };
 
-        events.forEach((event: MatrixEvent) => {
+        for (const event of events) {
```

Comment:
TIL that `forEach` is up to [3 times slower](https://leanylabs.com/blog/js-forEach-map-reduce-vs-for-for_of/#arrayforeach-vs-for-and-forof) than the classic loops. 😳

Reply:
Which makes me wonder if other noticeable performance improvements could be had from banning it elsewhere, too?

Reply:
I tested `forEach` vs. `for of` with their code in the browser.

Interesting results:

Firefox: 
forEach of took: 47
for of took: 77

Chrome:
forEach of took: 3807
for of took: 4095

<details>

<summary>Code</summary>

```
<!DOCTYPE html>
<html lang="de">
    <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>Titel</title>
    </head>
    <body>
        <div id="results">

        </div>

        <script>
            console.log("Starting");

            function generateTestArray() {
                const result = [];
                for (let i = 0; i < 10000000; ++i) {
                    result.push({
                        a: i,
                        b: i / 2,
                        r: 0,
                    });
                }
                return result;
            }

            const testArray1 = generateTestArray();
            const testArray2 = generateTestArray();

            const results = document.getElementById("results");

            const before1 = Date.now();
            testArray1.forEach((x) => {
                x.r = x.a + x.b;
            });
            const after1 = Date.now();

            const before2 = Date.now();
            for (const obj of testArray2) {
                obj.r = obj.a + obj.b;
            }
            const after2 = Date.now();

            const result1 = document.createElement("DIV");
            result1.textContent = "forEach of took: " + (after1 - before1);
            results.appendChild(result1);

            const result2 = document.createElement("DIV");
            result2.textContent = "for of took: " + (after2 - before2);
            results.appendChild(result2);
        </script>
    </body>
</html>
```

</details>

Reply:
Oh interesting. Somehow it makes no sense to me that these aren't optimized into the same machine code. 🤔

Reply:
Yet another benchmark: https://www.measurethat.net/Benchmarks/Show/18189/0/for-vs-foreach-vs-forin-vs-forof
This is more aimed at the overhead of the different ways of iterating an array.

I would conclude: The performance of a loop highly depends on what you do in it.

Reply:
It also depends on how the array is interpreted and what it contains.
I found https://v8.dev/blog/elements-kinds interesting to conceptualise some perf optimisation that happen at the JS engine level

I'd optimise for readibility in the source code, and if we found that one thing was really slower than other, looking at transpiling that at build time

Reply:
FTR the reason for the swap here was simply for access to async/await

Reply:
> Yet another benchmark: [measurethat.net/Benchmarks/Show/18189/0/for-vs-foreach-vs-forin-vs-forof](https://www.measurethat.net/Benchmarks/Show/18189/0/for-vs-foreach-vs-forin-vs-forof) This is more aimed at the overhead of the different ways of iterating an array.

That particular one showed a factor 100 difference between `for of` and `for in` in my browser. I find it mildly irritating that there's so much variation just from the choice of iteration mechanism.

> I'd optimise for readibility in the source code, and if we found that one thing was really slower than other, looking at transpiling that at build time

💯